### PR TITLE
[air] Tune to support ScalingConfigDataClass.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

Tune needs to support ScalingConfig as provided as a dataclass together with tune.grid_search/choice/uniform etc.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/impl/utils.py [hunk 1]
```diff
@@ -26,3 +31,18 @@ def execute_dataset(config_dict: dict):
...
+    it can be used to generate search space.
+    """
+    scaling_config = params_space.get("scaling_config", None)
+    if not isinstance(scaling_config, ScalingConfigDataClass):
+        return
+    scaling_config_dict = dict()
+    for field in dataclasses.fields(scaling_config):
+        attr = getattr(scaling_config, field.name)
+        if attr:
+            scaling_config_dict[field.name] = attr
```
Comment (krfricke) [inline python/ray/tune/impl/utils.py:47]:
```suggestion
    scaling_config_dict = scaling_config.__dict__.copy()
```
? :-)

**Review by krfricke — COMMENTED** [review]
Looks good! Minor nit to simplify conversion

**Review by krfricke — APPROVED** [review]

File: python/ray/tune/impl/utils.py [hunk 2]
```diff
@@ -26,3 +29,13 @@ def execute_dataset(config_dict: dict):
                     config_dict[k].categories = [
                         _item.fully_executed() for _item in _list
                     ]
+
+
+def process_scaling_config(params_space: dict):
+    """Convert ``params_space["scaling_config"]`` back to a dict so that
+    it can be used to generate search space.
+    """
+    scaling_config = params_space.get("scaling_config", None)
```
Comment (amogkam) [inline python/ray/tune/impl/utils.py:38]:
```suggestion
    scaling_config = params_space.get(_KEY_SCALING_CONFIG, None)
```

**Review by amogkam — COMMENTED** [review]
@xwjiang2010 will the actual API changes come in a follow up PR?

File: python/ray/tune/impl/utils.py [hunk 3]
```diff
@@ -26,3 +29,13 @@ def execute_dataset(config_dict: dict):
...
+    if not isinstance(scaling_config, ScalingConfigDataClass):
+        return
+    params_space[_KEY_SCALING_CONFIG] = scaling_config.__dict__.copy()
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/impl/utils.py:41]: hmmm, actually I think `.copy` is not even needed.
What is the concern here?

File: python/ray/tune/impl/utils.py [hunk 4]
```diff
@@ -1,14 +1,17 @@
 from ray.data import Dataset
+from ray.ml.config import ScalingConfigDataClass
 from ray.tune.sample import Categorical
 
+_KEY_SCALING_CONFIG = "scaling_config"
```
Comment (matthewdeng) [inline python/ray/tune/impl/utils.py:5]: nit

```suggestion
_SCALING_CONFIG_KEY = "scaling_config"
```

File: python/ray/tune/impl/utils.py [hunk 5]
```diff
@@ -26,3 +29,13 @@ def execute_dataset(config_dict: dict):
                     config_dict[k].categories = [
                         _item.fully_executed() for _item in _list
                     ]
+
+
+def process_scaling_config(params_space: dict):
+    """Convert ``params_space["scaling_config"]`` back to a dict so that
+    it can be used to generate search space.
+    """
+    scaling_config = params_space.get(_KEY_SCALING_CONFIG, None)
```
Comment (matthewdeng) [inline python/ray/tune/impl/utils.py:38]: Make the naming of `params_space` be consistent with `process_dataset_param(config_dict: dict)`?

File: python/ray/tune/impl/utils.py [hunk 6]
```diff
@@ -26,3 +29,13 @@ def execute_dataset(config_dict: dict):
                     config_dict[k].categories = [
                         _item.fully_executed() for _item in _list
                     ]
+
+
+def process_scaling_config(params_space: dict):
+    """Convert ``params_space["scaling_config"]`` back to a dict so that
+    it can be used to generate search space.
```
Comment (matthewdeng) [inline python/ray/tune/impl/utils.py:36]: Maybe I'm missing something obvious, but when/where does this get converted back to a `ScalingConfigDataClass` to be used by the `Trainer`??

Comment (xwjiang2010) [reply python/ray/tune/impl/utils.py:36]: Tuner doesn't interact with Trainer.__init__ directly does it? It only interacts with Trainable. So why not just keep this internal interface taking in a dict based ScalingConfig? I don't think we need to convert it back to dataclass. 

In short, Trainer speaks dataclass while Trainable speak dict.

Also to one of @amogkam's last comments, we could also update Trainer.__init__ to just take in dataclass in this PR.

File: python/ray/tune/impl/utils.py (hunk 5 above)
Comment (xwjiang2010) [reply python/ray/tune/impl/utils.py:38]: `process_dataset_param ` has a recursive impl, hence the naming.

**Comment by xwjiang2010** [comment]
To summarize my last discussion with @amogkam on this PR:
1. why we want to do it: because we want all the configs to be consistent, e.g. RunConfig etc.
2. what we don't quite like about this PR: ScalingConfig exposing interfaces like Domain (which is Tune specific). But we can't seem to find a way around it.
3. What is missing in this PR: What should happen on Trainer side. It's agreed that `Trainer.__init__` should take in ScalingConfigDataClass. The open question is in `as_trainable` method `config["scaling_config"]` should be a dataclass or just a dict. According to the current impl in this PR, `config["scaling_config"]` would be a dict. Either Tune side should convert it back to a dataclass before passing the `config` to `train_func` or `train_func` can deal with a dict based scaling config. 
We also talked about if Tune can just natively support any dataclass semantics somehow without special casing for just this dataclass.

**Comment by xwjiang2010** [comment]
not relevant anymore.

**Pull request closed without merging**